### PR TITLE
chore: disable sending updates to client for effectively non-visible nodes, and some additional cleanup, changes to apply CP-s (#15885)  (#16855) and (#16836) (CP: 1.0)

### PR DESCRIPTION
**Pull request by czp13**
## Description
Manually cherry-picked #15885 PR for 1.0:
- Disable sending updates to client for effectively non-visible nodes
- Fix test: Always set tested component visible first; only assume node has changed if testvalue differs from original value
- Cleanup, and refactoring, in Element, StateNode, UIInternals classes + mvn formatter.

Plus manually cherry-picked #16836 and #16855 PR-s commits, and fine-tuned for this codebase/branch.

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 1]
```diff
@@ -557,11 +564,18 @@ public List<JavaScriptInvocation> dumpPendingJavaScriptInvocations() {
             return Collections.emptyList();
         }
 
-        List<JavaScriptInvocation> currentList = pendingJsInvocations;
+        List<JavaScriptInvocation> readyToSend = getPendingJavaScriptInvocations()
+                .stream()
+                .filter(invocation -> invocation.getOwner().isPresent() && invocation.getOwner().get().isVisible())
+//                .peek(JavaScriptInvocation::setSentToBrowser)
```
Comment (czp13) [inline flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:570]: The only difference between the original and cherry-picked commit, this part can be deleted as this concept does not exist in `1.0` and is not used anywhere else in `1.0`. 

TLDR: So it is safe to delete the setting part (as getting, or using is not existing)

**Comment by mcollovati** [comment]
I think this manual cherry pick should also consider changes in #16836 and #16855

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 2]
```diff
@@ -557,11 +564,17 @@ public List<JavaScriptInvocation> dumpPendingJavaScriptInvocations() {
             return Collections.emptyList();
         }
 
-        List<JavaScriptInvocation> currentList = pendingJsInvocations;
+        List<JavaScriptInvocation> readyToSend = getPendingJavaScriptInvocations()
+                .stream()
+                .filter(invocation -> invocation.getOwner().isPresent() && invocation.getOwner().get().isVisible())
```
Comment (mshabarov) [inline flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:569]: Invocations that have no owners are ready to be send as well - we assume they are for root node, i.e. always visible.

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 3]
```diff
@@ -557,11 +564,17 @@ public List<JavaScriptInvocation> dumpPendingJavaScriptInvocations() {
...
+                .collect(Collectors.toList());
 
-        pendingJsInvocations = new ArrayList<>();
+        pendingJsInvocations = getPendingJavaScriptInvocations()
+                .stream()
+                .filter(invocation -> !(invocation.getOwner().isPresent() && invocation.getOwner().get().isVisible()))
```
Comment (mshabarov) [inline flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:574]: Same here: this filter condition needs change, because JS invocations for root node would be pending forever.

**Review by mshabarov — CHANGES_REQUESTED** [review]

**Comment by mshabarov** [comment]
> @mcollovati: I was thinking about merging this one soon.
> 
> And then merge the mentioned two commits together as the third (#16855) fix is a small one, but the second fix/CP (#16836) is quite big and building on top of this. It could be nice for historical reasons, reverting, git tree history to unite the PR-s, and commits like this.
> 
> p.s: but if you feel, I can add those ones to here 🙇

You can add the commits to this branch, but then do "Rebase & Merge", not "Squash" and they will be preserved in history.

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 4]
```diff
@@ -542,6 +543,10 @@ public ExecutionCanceler addJavaScriptInvocation(
             JavaScriptInvocation invocation) {
         session.checkHasLock();
         pendingJsInvocations.add(invocation);
+      invocation.getOwner().ifPresent(node -> invocation.getOwner().get()
+             .addDetachListener(() -> pendingJsInvocations
+                     .removeIf(pendingInvocation -> pendingInvocation
+                             .equals(invocation))));
```
Comment (mcollovati) [inline flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:549]: the `node` param in the lambda is already `invocation.getOwner().get()`
```suggestion
      invocation.getOwner().ifPresent(node -> 
              node.addDetachListener(() -> pendingJsInvocations
                     .removeIf(pendingInvocation -> pendingInvocation
                             .equals(invocation))));
```

Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:549]: Good one, shorter code is usually better. The only reason I do not commit this is I want to have only one commit for this change, then another commit for the other CP-s, to not push like 5-10 commits to the branch. 

So I squash logically related commits together.

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java (hunk 2 above)
Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java]: Thanks, fixed!

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 5]
```diff
@@ -557,11 +561,17 @@ public List<JavaScriptInvocation> dumpPendingJavaScriptInvocations() {
             return Collections.emptyList();
         }
 
-        List<JavaScriptInvocation> currentList = pendingJsInvocations;
+        List<JavaScriptInvocation> readyToSend = getPendingJavaScriptInvocations()
+                .stream()
+                .filter(invocation -> invocation.getOwner().isEmpty() || invocation.getOwner().get().isVisible())
```
Comment (mcollovati) [inline flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:566]:
```suggestion
                .filter(invocation -> invocation.getOwner().map(StateNode::isVisible).orElse(true))
```

[the remark this answers is not in the document]
Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:566]: Thanks, changed to negated isPresent() shall be the same.

Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:566]: This is good :), I've double-checked.

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 6]
```diff
@@ -557,11 +561,17 @@ public List<JavaScriptInvocation> dumpPendingJavaScriptInvocations() {
...
+                .collect(Collectors.toList());
 
-        pendingJsInvocations = new ArrayList<>();
+        pendingJsInvocations = getPendingJavaScriptInvocations()
+                .stream()
+                .filter(invocation -> invocation.getOwner().isPresent() && !invocation.getOwner().get().isVisible())
```
[the remark this answers is not in the document]
Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:571]: Ok, this seems the same by eye checks. 

Asked my bot friend, who said it is different, then I executed this code for test:

```

import lombok.Data;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class InvocationLogicTest {

    @Data
    static class Owner {
        private boolean isVisible;

        public Owner(boolean isVisible) {
            this.isVisible = isVisible;
        }

        public boolean isVisible() {
            return isVisible;
        }
    }

    @Data
    static class Invocation {
        private Owner owner;

        public Invocation(Owner owner) {
            this.owner = owner;
        }

        public Invocation() {
        }

        public Optional<Owner> getOwner() {
            return owner == null ? Optional.empty() : Optional.of(owner);
        }
    }

    public static void main(String[] args) {
        List<Invocation> invocations = List.of(
                new Invocation(new Owner(true)),
                new Invocation(),
                new Invocation(new Owner(false))
        );


        System.out.println("Logic for: invocations.stream().filter(invocation -> invocation.getOwner().isPresent() && !invocation.getOwner().get().isVisible()).collect(Collectors.toList())");
        List<Invocation> filteredList1 = invocations.stream()
                .filter(invocation -> invocation.getOwner().isPresent() && !invocation.getOwner().get().isVisible())
                .collect(Collectors.toList());

        System.out.println(filteredList1.size());  // Output: 1
        System.out.println(filteredList1.get(0) == invocations.get(2));  // Output: true
        System.out.println(filteredList1.get(0));

        System.out.println("Logic for: invocations.stream().filter(invocation -> invocation.getOwner().map(node -> !node.isVisible()).orElse(false)).collect(Collectors.toList())");
        List<Invocation> filteredList2 = invocations.stream()
                .filter(invocation -> invocation.getOwner().map(node -> !node.isVisible()).orElse(false))
                .collect(Collectors.toList());

        System.out.println(filteredList2.size());  // Output: 2
        System.out.println(filteredList2.get(0) == invocations.get(2));  // Output: true
        System.out.println(filteredList2.get(0));


    }

}

```

It showed this output:
```
Logic for: invocations.stream().filter(invocation -> invocation.getOwner().isPresent() && !invocation.getOwner().get().isVisible()).collect(Collectors.toList())
1
true
InvocationLogicTest.Invocation(owner=Optional[InvocationLogicTest.Owner(isVisible=false)])
Logic for: invocations.stream().filter(invocation -> invocation.getOwner().map(node -> !node.isVisible()).orElse(false)).collect(Collectors.toList())
1
true
InvocationLogicTest.Invocation(owner=Optional[InvocationLogicTest.Owner(isVisible=false)])
```

[the remark this answers is not in the document]
Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:571]: And then my bot friend asked for an excuse:

```
Apologies for the confusion caused by the previous response. Upon reviewing the code snippets, I realize that I made an error in explaining the difference. I apologize for any inconvenience caused.

In the given code snippets:

.filter(invocation -> invocation.getOwner().isPresent() && !invocation.getOwner().get().isVisible())
.filter(invocation -> invocation.getOwner().map(node -> !node.isVisible()).orElse(false))
Both snippets will produce the same output and behave identically in terms of filtering the invocations.

To clarify, in the second snippet, when invocation.getOwner() returns an empty optional (Optional.empty()), the map function will not be executed, and the orElse(false) part will immediately provide the default value of false. This means that invocations without an owner will be filtered out in both cases.

I apologize for the incorrect explanation in the previous response. Thank you for pointing it out, and I appreciate your understanding.
```

You can never believe anyone, even if it is a bot. Maybe too much effort but was interesting, I will push your suggestion because it was overtested as well.

[the remark this answers is not in the document]
Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:571]: Oh, yeah, and first it said it is different the 3 cases (he accidentally shown 4 cases, because duplicated one).

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 7]
```diff
@@ -16,16 +16,14 @@
...
 
 import java.io.Serializable;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
+import java.util.*;
```
Comment (czp13) [inline flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:19]: I will fix this, accidentally united.

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 8]
```diff
@@ -557,11 +558,93 @@ public List<JavaScriptInvocation> dumpPendingJavaScriptInvocations() {
...
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    private void registerDetachListenerForPendingInvocation(
+            JavaScriptInvocation invocation) {
+
+        StateNode ownerNode;
+        if (!invocation.getOwner().isPresent()) {
+//            return;
+            // or throw error?
+//            throw new IllegalStateException("Owner is not present for invocation: " + invocation);
+            ownerNode = ui.getInternals().getStateTree().getRootNode();
```
Comment (czp13) [inline flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:591]: Thinking it through:
1. if `invocation.getOwner()` is somehow is null should be equal with `getInternals().getStateTree().getRootNode()` (as we discussed before)
2. also the `registerDetachListenerForPendingInvocation` method should be called with the filled-out owner (and not visible rootnode) based on the 568. line logic

So I can remove those comments.

File: flow-server/src/test/java/com/vaadin/flow/component/internal/UIInternalsTest.java [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+//
+//        UIInternals.JavaScriptInvocation invocation =
+//                new UIInternals.JavaScriptInvocation("", node);
+//        internals.addJavaScriptInvocation(invocation);
+//        internals.dumpPendingJavaScriptInvocations();
+//
+//        Mockito.verify(node, Mockito.times(1))
+//                .addDetachListener(() -> {});
+//
+//        invocation.cancelExecution();
```
Comment (czp13) [inline flow-server/src/test/java/com/vaadin/flow/component/internal/UIInternalsTest.java:233]: no `cancelExecution`, nor `completeExceptionally`, `complete` methods are implemented for JavaScriptInvocation so I think this test can be deleted. (manipulating and setting up successHandler and errorHandler, which is also missing from this codebase)

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 10]
```diff
@@ -557,11 +564,80 @@ public List<JavaScriptInvocation> dumpPendingJavaScriptInvocations() {
...
+                            node.addDetachListener(detachListener));
+                    return detachListener;
+                });
+        listener.invocationList.add(invocation);
+
+        SerializableConsumer callback = unused -> listener
+                .onInvocationCompleted(invocation);
+        // not then part for JavaScript invocation, no successHandler or errorHandler:
+        // so we can remove this one too, I believe:
+        // invocation.then(callback, callback);
```
Comment (czp13) [inline flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:609]: `invocation.then` is the last method that is missing, but also related to `successHandler()` and `errorHandler()`, did not go that deep to see how much effort would be to add these to the code base in a separate PR.

File: flow-server/src/main/java/com/vaadin/flow/shared/Registration.java [hunk 11]
```diff
@@ -29,4 +31,20 @@ public interface Registration extends Serializable {
...
+     * @return a registration that removes all provided registrations not
+     *         <code>null</code>
+     * @since
+     */
+    static Registration combine(Registration... registrations) {
+        Objects.requireNonNull(registrations);
+        Arrays.asList(registrations).forEach(Objects::requireNonNull);
+
+        return () -> Arrays.asList(registrations).forEach(Registration::remove);
+    }
```
Comment (mcollovati) [inline flow-server/src/main/java/com/vaadin/flow/shared/Registration.java:49]: Probably it does not make sense to add this new API, that will be used only in one place.
It should be easy to apply this logic in the caller.

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java (hunk 10 above)
Comment (mcollovati) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:609]: Instead of adding the missing APIs, I wonder if we can simply call `onInvocationCompleted` in `dumpPendingJavaScriptInvocations`  for every invocation that will be sent to the browser (`readyToSend` list).
In this case, I would rename `onInvocationCompleted` to something that better explains that we are removing the detach listener because the invocation will be sent to the client and executed.
@mshabarov what do you think?

**Review by mcollovati — CHANGES_REQUESTED** [review]

File: flow-server/src/main/java/com/vaadin/flow/shared/Registration.java (hunk 11 above)
Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/shared/Registration.java:49]: Thanks Marco! 
My idea was that:
- this is a nice addition/improvement to the codebase, so it was a low hanging fruit, 
- and maybe also next time we will need it, we can find it here, but if it is a private method or part of a method then we will duplicate the code.

If you still feel I can refactor this part and do as you wish! 🙇

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java (hunk 10 above)
Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:609]: Thanks! Interesting idea,  I will wait for @mshabarov's thoughts on this, then I can check and do the necessary changes. 🙇

File: flow-server/src/main/java/com/vaadin/flow/shared/Registration.java (hunk 11 above)
Comment (mcollovati) [reply flow-server/src/main/java/com/vaadin/flow/shared/Registration.java:49]: My opinion is that we should not be ad new features to 1.0 if not strictly necessary.
In addition, if you want to peek at this a public API you should also port/implement related tests

Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/shared/Registration.java:49]: Adding tests, a very fair point! 👍  I have no problem adding more tests, other than the PR starting to be quite big and making it hard to review it properly (so maybe a separate PR would have been the best solution if we want to add it).

Adding new features to 1.0, but with this you bought me :D, indeed this branch should be just for security fixes, and absolutely necessary commits, and so not much point to add new API-s to it, as probably we will not use it, so yap, let's go with your approach.

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 12]
```diff
@@ -594,7 +596,7 @@ private void registerDetachListenerForPendingInvocation(
         JavaScriptInvocationDetachListener listener = pendingJsInvocationDetachListeners
                 .computeIfAbsent(ownerNode, node -> {
                     JavaScriptInvocationDetachListener detachListener = new JavaScriptInvocationDetachListener();
-                    detachListener.registration = Registration.combine(
+                    detachListener.registration = combine(
                             () -> pendingJsInvocationDetachListeners
                                     .remove(node),
                             node.addDetachListener(detachListener));
```
Comment (mcollovati) [inline flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java]: I would remove the `combine` method, as it introduces a lot of overhead for a single call (e.g, creating `List`s, as it expects potentially more than one registration)
Instead, I would simply create a lambda here

```suggestion
                    Registration registration = node.addDetachListener(detachListener);
                    detachListener.registration = () -> {
                        pendingJsInvocationDetachListeners.remove(node);
                        registration.remove();
                    };
```

**Review by mcollovati — CHANGES_REQUESTED** [review]

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 13]
```diff
@@ -557,11 +566,96 @@ public List<JavaScriptInvocation> dumpPendingJavaScriptInvocations() {
...
+        StateNode ownerNode = invocation.getOwner().orElse(ui.getInternals().getStateTree().getRootNode());
+
+        JavaScriptInvocationDetachListener listener = pendingJsInvocationDetachListeners
+                .computeIfAbsent(ownerNode, node -> {
+                    JavaScriptInvocationDetachListener detachListener = new JavaScriptInvocationDetachListener();
+                    detachListener.registration = combine(
+                            () -> pendingJsInvocationDetachListeners
+                                    .remove(node),
+                            node.addDetachListener(detachListener));
+                    return detachListener;
```
Comment (mcollovati) [inline flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:603]: I would remove the combine method, as it introduces a lot of overhead for a single call (e.g, creating Lists, as it expects potentially more than one registration)
Instead, I would simply create a lambda here
```suggestion
                    Registration registration = node.addDetachListener(detachListener);
                    detachListener.registration = () -> {
                        pendingJsInvocationDetachListeners.remove(node);
                        registration.remove();
                    };
```

(Sorry for duplicate, I messed up with comments)

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java (hunk 10 above)
Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:611]: I agree with this approach, but how can you retrieve `JavaScriptInvocationDetachListener` for `readyToSend` invocations? You need this listener to get the registration to call remove on it.

Ah...you already have a map with node vs listener. Let do it then.

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java (hunk 13 above)
Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java]: Thank you vm, makes sense, cleaner! Fixed it, but I could not accept the suggestion (will squash all last commits also no worries), because  `return detachListener;` was needed still.

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 14]
```diff
@@ -557,11 +567,84 @@ public List<JavaScriptInvocation> dumpPendingJavaScriptInvocations() {
...
+                .filter(invocation -> invocation.getOwner().map(StateNode::isVisible).orElse(true))
+                .collect(Collectors.toList());
+
+        pendingJsInvocationDetachListeners.entrySet().stream()
+                .flatMap(invocationListeners -> readyToSend.stream()
+                        .filter(invocation -> invocation.getOwner()
+                                .orElseGet(() -> ui.getInternals().getStateTree().getRootNode())
+                                .getId() == invocationListeners.getKey().getId()))
+                .forEach(invocation -> pendingJsInvocationDetachListeners.values()
+                        .forEach(listener -> listener.removePendingInvocation(invocation)));
```
Comment (mcollovati) [inline flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:577]: It seems a bit complicated to me. Also, it should call `listener.onInvocationCompleted()` instead of `listener.removePendingInvocation()` otherwise the invocation will not be removed from the `JavaScriptInvocationDetachListener`

And as commented before, method `onInvocationCompleted` should be renamed with something more related to the context where it is applied in this PR

```suggestion
        StateNode rootNode = ui.getInternals().getStateTree().getRootNode();
        readyToSend.forEach(i -> {
            JavaScriptInvocationDetachListener listener = pendingJsInvocationDetachListeners.get(i.getOwner().orElse(rootNode));
            if (listener != null) {
                listener.onInvocationCompleted(i);
            }
        });
```

**Review by mcollovati — CHANGES_REQUESTED** [review]

**Review by mcollovati — APPROVED** [review]

File: flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java [hunk 15]
```diff
@@ -557,11 +563,83 @@ public List<JavaScriptInvocation> dumpPendingJavaScriptInvocations() {
...
+            UIInternals.this.pendingJsInvocations.removeIf(
+                    pendingInvocation -> pendingInvocation.equals(invocation));
+            if (invocationList.isEmpty() && registration != null) {
+                registration.remove();
+                registration = null;
+            }
+        }
 
-        return currentList;
+        void removePendingStatusAndRegistrationAndDetachListener(JavaScriptInvocation invocation) {
```
[the remark this answers is not in the document]
Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:639]: Super! Thank you! 🎉 
Yes, not the best, I agree, but if the scope is narrower then the name can be longer, based on Uncle Bob clean coding paradigm :), so that is why I was brave enough to make it longer. (https://bpoplauschi.github.io/2021/01/30/Clean-Code-Naming-by-Uncle-Bob-part-3.html)

I will squash the commits and then we can merge! 🙇

[the remark this answers is not in the document]
Comment (mcollovati) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:639]: Let's also wait for @mshabarov

[the remark this answers is not in the document]
Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/component/internal/UIInternals.java:639]: Oki, sure. I squashed the commits for clearer `rebase and merge`.

**Review by mshabarov — APPROVED** [review]

**Comment by mshabarov** [comment]
@mcollovati if you have no comments, we can merge. @czp13 excellent job 👍

**Review by mcollovati — APPROVED** [review]
LGTM. @czp13 great job

**Comment by czp13** [comment]
Thank you very much for the constant reviews and ideas for both of you! 🙇
